File: run/peer_wrapper.R

#########################################
## Script to train a MOFA model from R ##
#########################################

# Important: a part from the MOFA2 R package, you also need the python package mofa2 to be installed
library(MOFA2)

###############
## Load data ##
###############

datafile <- "/Users/ricard/data/peer_wrapper/data.txt.gz"

# The data has to be loaded as a matrix with dimensions (features,samples)
data <- t( read.table(datafile) )

# MOFA is a multi-view factor analysis framework that is a generalisation of PEER
# The data needs to be input as a list of views. If you have a single data modality, the input data 
# corresponds to a list with a single element.
data <- list(data)

########################
## Create MOFA object ##
########################

object <- suppressMessages(create_mofa(data))

# check dimensionalities
print(object)

#########################
## Define data options ##
#########################

data_opts <- get_default_data_options(object)

##########################
## Define model options ##
##########################

model_opts <- get_default_model_options(object)

# use ARD prior for the factors? (please do not edit this)
model_opts$ard_factors <- FALSE

# number of factors
model_opts$num_factors <- 10


#############################
## Define training options ##
#############################

train_opts <- get_default_training_options(object)

# maximum number of iterations
train_opts$maxiter <- 2000

# fast, medium, slow
train_opts$convergence_mode <- "fast"

# initial iteration to start evaluating convergence using the ELBO (recommended >1)
train_opts$startELBO <- 1

# frequency of evaluation of ELBO (recommended >1)
train_opts$freqELBO <- 1

# use GPU (needs cupy installed and working)
train_opts$gpu_mode <- FALSE

# verbose output?
train_opts$verbose <- FALSE

# random seed
train_opts$seed <- 1

#############################
## Prepare the MOFA object ##
#############################

object <- prepare_mofa(
  object = object, 
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)

##############
## Run MOFA ##
##############

# The output is an hdf5 file that can be loaded using the function load_model(file)
outfile = "/Users/ricard/data/peer_wrapper/model.hdf5"
model <- run_mofa(object, outfile=outfile)

File: run/peer_wrapper.py
######################################################
## Template script to train a MOFA model via Python ##
######################################################

from mofapy2.run.entry_point import entry_point
import numpy as np
import pandas as pd

###############
## Load data ##
###############

datafile = "/Users/ricard/data/peer_wrapper/data.txt.gz"

# The data has to be loaded as a pandas dataframe or as a numpy matrix with dimensions (samples,features)
data = pd.read_csv(datafile, header=0, sep='\t')

###########################
## Initialise MOFA model ##
###########################

# initialise the entry point
ent = entry_point()

# Set data
# MOFA is a multi-view and multi-group inference framework. 
# If usig only a single view and a single group (as in PEER), the data needs to be embedded into a nested list
ent.set_data_matrix([[data]])

# Set model options
# - factors: number of factors
# - spikeslab_weights: use spike-and-slab sparsity on the loading?
# - ard_weights: use ARD prior on the loadings (please do not edit this)
ent.set_model_options(factors=10, spikeslab_weights=False, ard_weights=False)

# Set training options
# - iter: maximum number of iterations
# - convergence_mode: fast, medium, slow
# - startELBO: initial iteration to start evaluating convergence using the ELBO (recommended >1)
# - freqELBO: frequency of evaluation of ELBO (recommended >1)
# - gpu_mode: use GPU (need cupy installed and working)
# - verbose: verbosity
# - seed
ent.set_train_options(iter=1000, convergence_mode="fast", startELBO=1, freqELBO=1, gpu_mode=False, verbose=False, seed=42)

####################################
## Build and train the MOFA model ##
####################################

ent.build()
ent.run()

####################
## Save the model ##
####################

# The output is an hdf5 file that can be subsequently loaded into R (no python package for the downstream analysis is available yet)
outfile = "/Users/ricard/data/peer_wrapper/model.hdf5"
ent.save(outfile)
